### PR TITLE
Add support for tailable cursors

### LINKED ISSUE #2283: Support tailable cursors
### What should be done?

https://www.mongodb.com/docs/manual/core/tailable-cursors/

Tailable cursors do not close when it exhausted results, the cursor is kept open waiting for more documents to be added. It's available for find cursor. Not available for aggregate cursor.

Support both `tailable` and `awaitData`.

```[tasklist]
### Tasks
- [ ] Support `awaitData`
- [x] Check test coverage, add tests if needed
- [x] Clean-up implementation
```

### Where?

TBD

### Definition of Done

TBD

- all handlers updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated or an issue to create documentation created;

### PR DESCRIPTION
**Pull request by AlekSi**
# Description

Closes #2283.

To be merged on Monday after v1.17 release.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [x] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [x] I updated user documentation.
- [x] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/clientconn/cursor/cursor.go [hunk 1]
```diff
@@ -67,30 +98,81 @@ func newCursor(id int64, iter types.DocumentsIterator, params *NewParams, r *Reg
...
+	}
 
-		close(c.closed)
+	c.l.Debug("Closing cursor's iterator")
+	c.iter.Close()
+	c.iter = nil
+
+	c.m.Unlock()
+
+	// It is not entirely clear if we should do that; more tests are needed.
```
Comment (AlekSi) [inline internal/clientconn/cursor/cursor.go:170]: @noisersup, do we have tests checking that normal cursors disappear after they are exhausted?

File: internal/handler/msg_find.go [hunk 2]
```diff
@@ -128,41 +126,57 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
...
+	if params.SingleBatch || len(docs) < int(params.BatchSize) {
+		c.Close()
 
-	if params.SingleBatch || firstBatch.Len() < int(params.BatchSize) {
-		// support tailable cursors
-		// TODO https://github.com/FerretDB/FerretDB/issues/2283
+		// It is not entirely clear if we should do that; more tests are needed.
+		if c.Type != cursor.Normal {
+			h.cursors.CloseAndRemove(c)
+		}
```
Comment (AlekSi) [inline internal/handler/msg_find.go:171]: Hm…

Comment (AlekSi) [reply internal/handler/msg_find.go:171]: All tests also pass if I change `!=` to `==`, so yeah, we need to add a test for that

File: internal/clientconn/cursor/cursor.go [hunk 3]
```diff
@@ -27,38 +27,69 @@ import (
...
-	closed    chan struct{}
-	ID        int64
-	closeOnce sync.Once
+	r            *Registry
+	l            *zap.Logger
+	token        *resource.Token
+	removed      chan struct{} // protected by m
+	ID           int64
+	lastRecordID int64 // protected by m
+	m            sync.Mutex
```
Comment (henvic) [inline internal/clientconn/cursor/cursor.go:76]: If they group nicely together you might want to consider this (if you're not worried about data alignment):

```suggestion
	m            sync.Mutex // protects the following fields
	iter         types.DocumentsIterator
	removed      chan struct{}
	lastRecordID int64
```

**Review by henvic — APPROVED** [review]

File: internal/clientconn/cursor/cursor.go (hunk 3 above)
Comment (AlekSi) [reply internal/clientconn/cursor/cursor.go:76]: We have many `Cursor` instances, so yeah, here we care about values being smaller

**Review by henvic — APPROVED** [review]

File: internal/clientconn/cursor/cursor.go (hunk 1 above)
Comment (noisersup) [reply internal/clientconn/cursor/cursor.go:170]: Yep `TestCursorGetMoreCommand`/`GetMoreCursorExhausted`.

File: internal/handler/msg_find.go (hunk 2 above)
Comment (noisersup) [reply internal/handler/msg_find.go:171]: I've added this one: https://github.com/FerretDB/FerretDB/pull/3859/files#diff-ac456a92d3f0b992305fd1927bcc41d7c401a1d4622bcaddbee6e597605761e9R248 

It's still not dependent on this line, although I couldn't find any way to test it. Find won't return you any cursorID so you're not able to track it

**Review by rumyantseva — COMMENTED** [review]
Ok, I tried some things, and I don't see obvious problems. I know @noisersup might have some ideas, so his review is crucial for this PR.

File: internal/handler/common/find.go [hunk 4]
```diff
@@ -79,5 +79,7 @@ func GetFindParams(doc *types.Document, l *zap.Logger) (*FindParams, error) {
 		return nil, err
 	}
 
+	// check AwaitData=true, Tailable=false?
```
Comment (noisersup) [inline internal/handler/common/find.go:82]: It should return the error

**Review by noisersup — APPROVED** [review]
Everything seems good, although I'm still investigating how does AwaitData cursors behave. I'd suggest to potentailly fix in seperate PR after test are merged

**Pull request merged**